STYLE_GUIDE.md

```diff
@@ -0,0 +1,48 @@
+# Style guide
+
+This style guide and defines specific coding standards and advice for this Java codebase. The rules here are extensions to the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html).
+
+Please see the [contributing guide](CONTRIBUTING.md) for general guidance for contributing to this project.
+
+### Automatic formatting
+
+Automatic formatting should be performed with `./gradlew goJF` or `./mvnw fmt:format`.
+
+### Class member order
+
+*Extends [3.4.2](https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents)*
+
+Class members should in the following order in decreasing priority:
+
+1. Static before non-static
+1. Nested classes/interfaces before fields before constructors before methods
+1. Public before private
+1. Final before non-final
+
+### Public APIs
+
+User-facing methods (such as those in Jib Core) should not have types in their signature that are not standard JDK classes. For example, a parameter should take type `List` rather than Guava's `ImmutableList`.
+
+### Package hierarchy
+
+Packages should depend on each other without cycles.
+
+The following is a list of current `jib-core` packages (under `com.google.cloud.tools.jib`) and their immediate dependencies. These can be amended as code changes, but there should not be cyclical dependencies.
+
+- `api`
```

Comment:
This section is a bit hard to read, I wonder if we could somehow arrange this in a tree/pyramid structure to make the relationships more clear? Maybe not possible until we fix the cyclical dependencies.

Reply:
Yea, I think we can iterate on this to make it more readable as we fix the cycles.

Reply:
Could you use `dot` or something to express this as a graph?

Reply:
It's a graph generator from [GraphViz](https://www.graphviz.org).  There's a [web generator](http://www.webgraphviz.com).

Reply:
Oh cool, I'll try that out.

Reply:
Oh man I tried it out but the graph is quite messy:

```
digraph G {
"api"
"async"
"blob" -> "filesystem"
"blob" -> "hash"
"blob" -> "image" [label = "cycle - should fix" ]
"builder" -> "async"
"builder" -> "blob"
"builder" -> "cache"
"builder" -> "configuration"
"builder" -> "docker"
"builder" -> "event"
"builder" -> "filesystem"
"builder" -> "global"
"builder" -> "http"
"builder" -> "image"
"builder" -> "json"
"builder" -> "registry"
"cache" -> "blob"
"cache" -> "filesystem"
"cache" -> "hash"
"cache" -> "image"
"cache" -> "json"
"configuration" -> "cache"
"configuration" -> "filesystem"
"configuration" -> "event"
"configuration" -> "image"
"configuration" -> "registry"
"docker" -> "blob"
"docker" -> "cache"
"docker" -> "image"
"docker" -> "json"
"docker" -> "tar"
"event"
"filesystem"
"frontend" -> "configuration"
"frontend" -> "event"
"frontend" -> "filesystem"
"frontend" -> "image"
"frontend" -> "registry"
"global"
"hash" -> "blob"
"hash" -> "image"
"http" -> "blob"
"image" -> "blob"
"image" -> "configuration" [ label = "cycle - should fix - ImageToJsonTranslator" ]
"image" -> "filesystem"
"image" -> "json"
"image" -> "tar"
"json" -> "blob"
"registry" -> "blob"
"registry" -> "builder" [ label = "cycle - should fix - RegistryClient" ]
"registry" -> "configuration" [ label = "cycle - should fix - DockerConfigCredentialRetriever" ]
"registry" -> "event"
"registry" -> "global"
"registry" -> "http"
"registry" -> "image"
"registry" -> "json"
"tar" -> "blob"
}
```
resulting graph:
![graph](https://i.imgur.com/ESXmXOu.png)